chain/pruned_block_dispatcher.go

```diff
@@ -0,0 +1,554 @@
+			// querying the peer for blocks.
+			OnVerAck: func(*peer.Peer, *wire.MsgVerAck) {
+				close(ready)
+			},
+			// Register a callback to signal us whenever the peer
+			// has sent us a block message.
+			OnRead: func(p *peer.Peer, _ int, msg wire.Message, _ error) {
+				var block *wire.MsgBlock
+				switch msg := msg.(type) {
+				case *wire.MsgBlock:
+					block = msg
+				case *wire.MsgVersion, *wire.MsgVerAck:
+					return
+				default:
+					log.Debugf("Received unexpected message "+
+						"%T from peer %v", msg, p.Addr())
+					return
+				}
+
+				select {
+				case msgsRecvd <- block:
+				case <-d.quit:
+				}
+			},
+		},
+		AllowSelfConns: true,
+	}
+	p, err := peer.NewOutboundPeer(cfg, peerInfo.Addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &queryPeer{
+		Peer:      p,
+		ready:     ready,
+		msgsRecvd: msgsRecvd,
+		quit:      make(chan struct{}),
+	}, nil
+}
+
+// connectToPeer attempts to establish a connection to the given peer and waits
+// up to PeerReadyTimeout for the version exchange to complete so that we can
+// begin sending it our queries.
+func (d *PrunedBlockDispatcher) connectToPeer(peer *queryPeer) error {
+	conn, err := d.cfg.Dial(peer.Addr())
+	if err != nil {
+		return err
+	}
+	peer.AssociateConnection(conn)
+
+	select {
+	case <-peer.ready:
+	case <-time.After(d.cfg.PeerReadyTimeout):
+		peer.Disconnect()
+		return errors.New("timed out waiting for protocol negotiation")
+	case <-d.quit:
+		return errors.New("shutting down")
+	}
+
+	// Remove the peer once it has disconnected.
+	peer.signalUponDisconnect(func() {
+		d.peerMtx.Lock()
+		delete(d.currentPeers, peer.Addr())
+		d.peerMtx.Unlock()
+	})
+
+	return nil
+}
+
+// Query submits a request to query the information of the given blocks.
+func (d *PrunedBlockDispatcher) Query(blocks []*chainhash.Hash,
+	opts ...query.QueryOption) (<-chan *wire.MsgBlock, <-chan error) {
+
+	reqs, blockChan, err := d.newRequest(blocks)
+	if err != nil {
+		errChan := make(chan error, 1)
+		errChan <- err
+		return nil, errChan
+	}
+
+	var errChan chan error
+	if len(reqs) > 0 {
+		errChan = d.workManager.Query(reqs, opts...)
+	}
+	return blockChan, errChan
+}
+
+// newRequest construct a new query request for the given blocks to submit to
+// the internal workManager. A channel is also returned through which the
+// requested blocks are sent through.
+func (d *PrunedBlockDispatcher) newRequest(blocks []*chainhash.Hash) (
+	[]*query.Request, <-chan *wire.MsgBlock, error) {
+
+	// Make sure the channel is buffered enough to handle all blocks.
+	blockChan := make(chan *wire.MsgBlock, len(blocks))
+
+	d.blockMtx.Lock()
+	defer d.blockMtx.Unlock()
+
+	// Each GetData message can only include up to wire.MaxInvPerMsg invs,
+	// and each block consumes a single inv.
+	var reqs []*query.Request
+	var getData *wire.MsgGetData
+	for i, block := range blocks {
+		if getData == nil {
+			getData = wire.NewMsgGetData()
+		}
+
+		if _, ok := d.blocksQueried[*block]; !ok {
```

Comment:
What about adding a `debug` log here and below so we can track the number of concurrent requests for a single block in practice when this is active? So one here when we need a new query, and another when we just need to append the block channel as there's already an active query for that hash.

Reply:
Done, let me know if that's what you had in mind.